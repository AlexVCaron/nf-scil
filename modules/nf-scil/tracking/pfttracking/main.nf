
process TRACKING_PFTTRACKING {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_1.6.0.sif':
        'scilus/scilus:1.6.0' }"

    input:
        tuple val(meta), path(fodf), path(seed), path(include), path(exclude)

    output:
        tuple val(meta), path("*__pft_tracking.trk")            , emit: trk
        path "versions.yml"                                     , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def pft_random_seed = task.ext.pft_random_seed ?: "--seed " + task.ext.pft_random_seed : ""
    def compress = task.ext.pft_compress_streamlines ? "--compress " + task.ext.pft_compress_value : ""
    def pft_algo = task.ext.pft_algo ? "--algo " + task.ext.pft_algo: ""
    def pft_seeding_type = task.ext.pft_seeding ? "--"  + task.ext.pft_seeding : ""
    def pft_nbr_seeds = task.ext.pft_nbr_seeds ? "-- "  + task.ext.pft_nbr_seeds : ""
    def pft_step = task.ext.pft_step ? "--step "  + task.ext.pft_step : ""
    def pft_seeding_mask = task.ext.pft_seeding_mask_type ? "--"  + task.ext.pft_seeding_mask_type : ""
    def pft_theta = task.ext.pft_theta ? "--theta "  + task.ext.pft_theta : ""
    def pft_sfthres = task.ext.pft_sfthres ? "--sfthres "  + task.ext.pft_sfthres : ""
    def pft_sfthres_init = task.ext.pft_sfthres_init ? "--sfthres_init "  + task.ext.pft_sfthres_init : ""
    def pft_min_len = task.ext.pft_min_len ? "--min_length "  + task.ext.pft_min_len : ""
    def pft_max_len = task.ext.pft_max_len ? "--max_length "  + task.ext.pft_max_len : ""
    def pft_particles = task.ext.pft_particles ? "--particles "  + task.ext.pft_particles : ""
    def pft_back = task.ext.pft_back ? "--back "  + task.ext.pft_back : ""
    def pft_front = task.ext.pft_front ? "--forward "  + task.ext.pft_front : ""
    def basis = task.ext.basis ? "--sh_basis "  + task.ext.basis : ""

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1
    echo $pft_seeding_type
    scil_compute_pft.py $fodf $seed $include $exclude tmp.trk\
        $pft_algo $pft_seeding_type $pft_nbr_seeds \
        $pft_random_seed $pft_step $pft_theta\
        $pft_sfthres $pft_sfthres_init\
        $pft_min_len $pft_max_len\
        $pft_particles $pft_back\
        $pft_front $compress $basis

    scil_remove_invalid_streamlines.py tmp.trk\
        ${prefix}__pft_tracking.trk\
        --remove_single_point

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 1.6.0
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    
    """
    scil_compute_pft.py -h
    scil_remove_invalid_streamlines.py -h

    touch ${prefix}__pft_tracking.trk

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 1.6.0
    END_VERSIONS
    """
}
