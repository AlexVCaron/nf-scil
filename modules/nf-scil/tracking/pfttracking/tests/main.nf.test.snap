{
    "tracking - pfttracking - fa": {
        "content": [
            "test__pft_tracking.trk",
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__pft_tracking_config.json:md5,d8a7d811f0c9bec571e8e583b0aecb94"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__map_include.nii.gz:md5,b974759b87b8d920195d93f86219f2ee"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__map_exclude.nii.gz:md5,2876363d0394ea01534738650d78f441"
                ]
            ],
            [
                "versions.yml:md5,f0a7414d398c9d522071425143a92959"
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__pft_seeding_mask.nii.gz:md5,fd41bd41399bd7234721715e6ecca018"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-08T19:52:48.63142"
    },
    "tracking - pfttracking - interface": {
        "content": [
            "test__pft_tracking.trk",
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__pft_tracking_config.json:md5,7e326ef48dfda3d63e288f3f49d29847"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__map_include.nii.gz:md5,b974759b87b8d920195d93f86219f2ee"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__map_exclude.nii.gz:md5,2876363d0394ea01534738650d78f441"
                ]
            ],
            [
                "versions.yml:md5,f0a7414d398c9d522071425143a92959"
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__pft_seeding_mask.nii.gz:md5,4c9bb24b937fa8ee77923877f2ee6fd1"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-08T19:53:06.317605"
    },
    "tracking - pfttracking - wm": {
        "content": [
            "test__pft_tracking.trk",
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__pft_tracking_config.json:md5,bedf3ac8834d78389b3b58809f0bb750"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__map_include.nii.gz:md5,b974759b87b8d920195d93f86219f2ee"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__map_exclude.nii.gz:md5,2876363d0394ea01534738650d78f441"
                ]
            ],
            [
                "versions.yml:md5,f0a7414d398c9d522071425143a92959"
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__pft_seeding_mask.nii.gz:md5,4710e63db72f7dda7c10e46346bf98d3"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-08T19:52:30.179614"
    }
}